internal/databroker/config_source.go

```diff
@@ -0,0 +1,233 @@
+package databroker
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/pomerium/pomerium/config"
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/pkg/grpc"
+	configpb "github.com/pomerium/pomerium/pkg/grpc/config"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+)
+
+var (
+	configTypeURL string
+)
+
+func init() {
+	any, _ := ptypes.MarshalAny(new(configpb.Config))
+	configTypeURL = any.GetTypeUrl()
+}
+
+// ConfigSource provides a new Config source that decorates an underlying config with
+// configuration derived from the data broker.
+type ConfigSource struct {
+	mu               sync.RWMutex
+	computedConfig   *config.Config
+	underlyingConfig *config.Config
+	dbConfigs        map[string]*configpb.Config
+	updaterHash      uint64
+	cancel           func()
+	serverVersion    string
+	recordVersion    string
+
+	config.ChangeDispatcher
+}
+
+// NewConfigSource creates a new ConfigSource.
+func NewConfigSource(underlying config.Source, listeners ...config.ChangeListener) *ConfigSource {
+	src := &ConfigSource{
+		dbConfigs: map[string]*configpb.Config{},
+	}
+	for _, li := range listeners {
+		src.OnConfigChange(li)
+	}
+	underlying.OnConfigChange(func(cfg *config.Config) {
+		src.mu.Lock()
+		src.underlyingConfig = cfg.Clone()
+		src.mu.Unlock()
+
+		src.rebuild(false)
+	})
+	src.underlyingConfig = underlying.GetConfig()
+	src.rebuild(true)
+	return src
+}
+
+// GetConfig gets the current config.
+func (src *ConfigSource) GetConfig() *config.Config {
+	src.mu.RLock()
+	defer src.mu.RUnlock()
+
+	return src.computedConfig
+}
+
+func (src *ConfigSource) rebuild(firstTime bool) {
+	src.mu.Lock()
+	defer src.mu.Unlock()
+
+	cfg := src.underlyingConfig.Clone()
+	defer func() {
+		src.computedConfig = cfg
+		if !firstTime {
```

Comment:
Could we use `sync.Once` here instead?

Reply:
This is doing the opposite of sync.Once. It runs every time except the first time.

Reply:
@calebdoxsey Ah right 👍